### PR TITLE
Show Celery stats in webui

### LINKED ISSUE #11623: Add celery stats to Airflow webui
**Description**

Show basic information about Celery workers in Airflow webui.

**Use case / motivation**

Currently, users using CeleryExecutor have to use Flower to access information about Celery workers. I would like to propose adding new view in Airflow webui that would show similar information.

This should be easily doable by using:
```py
from airflow.executors.celery_executor import app

inspect = self.app.control.inspect(timeout=0.01, destination=None)
result = partial(getattr(inspect, "stats"))()
```

Note: this is how Flower gets the information.

**Related Issues**

N/A

### PR DESCRIPTION
**Pull request by turbaszek**
Add simple view showing information about Celery workers.
Closes: #11623

Screenshot 2020-10-18 at 18 48 17

~~I'm still facing some issue with the stability of getting stats (some missing keys and redis errors like here https://github.com/celery/kombu/issues/1063)~~

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by turbaszek** [comment]
> I'm still facing some issue with the stability of getting stats (some missing keys and redis errors like here celery/kombu#1063)

I think increasing timeout in inspect solved this issue 👍

**Comment by turbaszek** [comment]
CC @ryanahamilton @olchas

File: airflow/www/extensions/init_views.py
Comment (kurtqq) [inline airflow/www/extensions/init_views.py]: shouldn't be also a permission that can be granted to access this page for custom roles?

Comment (turbaszek) [reply airflow/www/extensions/init_views.py]: @kurtqq good catch, I will add

File: airflow/www/views.py
Comment (ashb) [inline airflow/www/views.py]: I think this should be a method on the executor -- this lets us add the capability for stats to other executors in the future.

File: airflow/www/views.py
Comment (ashb) [inline airflow/www/views.py:2642]:
```suggestion
            'airflow/celery_executor_stats.html',
```

File: airflow/www/extensions/init_views.py
Comment (ashb) [inline airflow/www/extensions/init_views.py]:
```suggestion
    appbuilder.add_view(views.StatsView, "Executor Stats", category="Admin")
```

File: airflow/utils/celery.py
Comment (ashb) [inline airflow/utils/celery.py]: I think all this should go in to the celery_executor.py (either directly as methods, or in a nested class under CeleryExecutor)

My main reason is that anything under `utils/` is a code-smell to me - it quickly becomes a junkyard where we just place anything and everything :)

File: airflow/www/extensions/init_views.py
Comment (ashb) [inline airflow/www/extensions/init_views.py]: Does it also make sense to show this if not on Celery/CeleryKube executor?

File: airflow/www/views.py
Comment (ashb) [reply airflow/www/views.py:2642]: Which could possibly be template which extends the common `stats.html` template.

File: airflow/utils/celery.py
Comment (ashb) [inline airflow/utils/celery.py]: What does the `partial` here do? Wouldn't `getattr(inspect, method)()`  do the same?

File: airflow/www/templates/airflow/stats.html
Comment (ashb) [inline airflow/www/templates/airflow/stats.html:22]: We should override the `page_title` too so we set the `<title>`.

File: airflow/www/templates/airflow/stats.html
Comment (ashb) [inline airflow/www/templates/airflow/stats.html]:
```suggestion
    <h2>State Summary</h2>
```

File: airflow/www/views.py
Comment (ashb) [inline airflow/www/views.py]: Not used by anything?

**Review by ashb — CHANGES_REQUESTED** [review]
Nice but lets make it a little bit less Celery specific.

File: airflow/utils/celery.py
Comment (turbaszek) [reply airflow/utils/celery.py]: I think it should but I took this cod from flower 😄

File: airflow/www/extensions/init_views.py
Comment (turbaszek) [reply airflow/www/extensions/init_views.py]: @ashb I added information about task instances and dag runs. When users use something else than celery executor then they will only see this simple info.

File: airflow/utils/celery.py
Comment (turbaszek) [reply airflow/utils/celery.py]: Happy to do that, I was not sure where to put this

Comment (ashb) [reply airflow/utils/celery.py]: If we do https://github.com/apache/airflow/pull/11631#discussion_r507568576 -- it could live somewhere in celery_executor.py

File: airflow/www/extensions/init_views.py
Comment (mik-laj) [reply airflow/www/extensions/init_views.py]: It would be great if we could also get information from Kubernetes. We can do it separately, but the method's interface should be generic enough to be useful for KubernetesExecutor as well.

File: airflow/www/templates/airflow/stats.html
Comment (ryanahamilton) [inline airflow/www/templates/airflow/stats.html]:
```suggestion
     <table class="table table-bordered table-striped" style="width: auto; min-width: 50%;">
```

- The `id` isn't necessary
- Probably not a big deal with this limited content, but setting the `width: 50%` can yield some undesirable effects with narrower screen-widths. This `width/min-width` combo will be a bit more flexible.

File: airflow/www/templates/airflow/stats.html
Comment (ryanahamilton) [inline airflow/www/templates/airflow/stats.html]: This top-level `<div>` is superfluous, since all of the contained children are block elements.

File: airflow/www/templates/airflow/stats.html
Comment (ryanahamilton) [inline airflow/www/templates/airflow/stats.html]:
```suggestion
     <table class="table table-bordered table-striped" style="width: auto; min-width: 50%;">
```
The `id` isn't necessary

File: airflow/www/templates/airflow/stats.html
Comment (ryanahamilton) [inline airflow/www/templates/airflow/stats.html]:
```suggestion
    <table class="table table-bordered table-striped">
```
The `id` isn't necessary

File: airflow/www/extensions/init_views.py
Comment (ashb) [reply airflow/www/extensions/init_views.py]: Yeah, that's exactly what I was thinking.

File: airflow/www/templates/airflow/stats.html
Comment (ashb) [reply airflow/www/templates/airflow/stats.html]: Trailing `;` in your suggestion - on purpose?

Comment (ryanahamilton) [reply airflow/www/templates/airflow/stats.html]: I know it's not required for the last one, but I like to add them for consistency sake. (If this weren't inline CSS, the stylelint rules would require it.)

File: airflow/www/views.py
Comment (turbaszek) [reply airflow/www/views.py:2642]: To be honest I'm not sure if I see value of adding more templates

File: airflow/www/templates/airflow/stats.html
Comment (mik-laj) [inline airflow/www/templates/airflow/stats.html:69]: Can you use [get_docs_link](https://github.com/apache/airflow/blob/master/airflow/utils/docs.py) function to generate link to fixed version?

File: airflow/www/templates/airflow/stats.html
Comment (ryanahamilton) [inline airflow/www/templates/airflow/stats.html:29]:
```suggestion
 <table class="table table-bordered table-striped" style="width: auto; min-width: 50%;">
```
Typo

File: airflow/www/templates/airflow/stats.html
Comment (ryanahamilton) [inline airflow/www/templates/airflow/stats.html:49]: Looks like this previous suggestion got lost:
```suggestion
 <table class="table table-bordered table-striped" style="width: auto; min-width: 50%;">
```

**Comment by kaxil** [comment]
Can you please rebase your PR on latest Master since we have applied [Black](https://github.com/apache/airflow/commit/4e8f9cc8d02b29c325b8a5a76b4837671bdf5f68) and [PyUpgrade](https://github.com/apache/airflow/commit/8c42cf1b00c90f0d7f11b8a3a455381de8e003c5) on Master.

It will help if your squash your commits into single commit first so that there are less conflicts.

**Comment by HaloKo4** [comment]
@turbaszek this looks amazing! is this part of Airflow 2 ?

**Comment by turbaszek** [comment]
> @turbaszek this looks amazing! is this part of Airflow 2 ?

I think we can do it, however it's not necessary. @kaxil @potiuk should we include this or should we limit the scope?

**Comment by potiuk** [comment]
If nothing else will be blocked by this one - why not. Seems rather disconnected from everything else and it has a very low risk. However I think at most we should rebase it now and only include if we manage to properly review it (after implementing all the high priority stuff)

**Comment by kaxil** [comment]
Yeah, let's only spend time on it after other high-priority items are done. And once those are done, I have no problems merging this in :)

**Comment by HaloKo4** [comment]
Rechecking about this. Is it expected to be included in 2.1?

**Pull request closed without merging**
